File: go/0002-add-two-numbers.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var dummy = new(ListNode)
	var l3 **ListNode = &dummy
	var carry int
	list1 := l1
	list2 := l2

	for (list1 != nil) || (list2 != nil) {
		l3 = &((*l3).Next)
		*l3 = new(ListNode)
		var sum int

		if list1 != nil {
			sum += list1.Val
			list1 = list1.Next
		}

		if list2 != nil {
			sum += list2.Val
			list2 = list2.Next
		}

		(*l3).Val = (sum + carry) % 10
		carry = (sum + carry) / 10
	}

	if carry > 0 {
		l3 = &((*l3).Next)
		*l3 = new(ListNode)
		(*l3).Val = carry
	}

	return dummy.Next
}

File: rust/0002-add-two-numbers.rs
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut l1, mut l2) = (l1.as_ref(), l2.as_ref());
        let mut dummy = Box::new(ListNode { val: 0, next: None });
        let mut cur = dummy.as_mut();
        let mut carry = 0;
    
        while l1.is_some() || l2.is_some() || carry > 0 {
            let val1 = l1.map(|n| n.val).unwrap_or_default();
            let val2 = l2.map(|n| n.val).unwrap_or_default();
    
            let sum = val1 + val2 + carry;
            carry = sum / 10;
            cur.next = Some(Box::new(ListNode { val: sum % 10, next:  None }));
            cur = cur.next.as_mut().unwrap();
    
            if let Some(node) = l1 { l1 = node.next.as_ref() }
            if let Some(node) = l2 { l2 = node.next.as_ref() }
        }
    
        dummy.next
    }
}
